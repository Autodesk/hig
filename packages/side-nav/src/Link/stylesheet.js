export default function(props, themeData) {
  return {
    wrapper: {
      display: "flex",
      justifyContent: "space-between",
      alignItems: "center",
      textDecoration: "none",
      padding: `0 ${themeData["density.spacings.medium"]}`,

      // we can't override the :hover selector on typography's
      // style prop, so we have to style the text from the
      // container element
      "&:hover span": {
        color: themeData["colorScheme.accentColor"]
      },

      // making sure to consitently use pointer cursor when
      // just implementing onClick and not setting link property
      "&:hover": {
        cursor: "pointer"
      },

      "&:hover svg *": {
        fill: themeData["colorScheme.accentColor"]
      }
    },

    externalIcon: {
      display: "flex",
      alignItems: "center",
      marginLeft: "auto",
      padding: "6px 2px"
    },

    typography: {
      lineHeight: "26px"
    }
  };
}
