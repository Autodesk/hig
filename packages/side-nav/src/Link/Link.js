import React, { Component } from "react";
import PropTypes from "prop-types";
import { sizes as iconSizes } from "@hig/icons";
import ThemeContext from "@hig/theme-context";
import Typography from "@hig/typography";
import { css } from "emotion";

import { AVAILABLE_TARGETS } from "../targets";
import ExternalLinkIcon from "../presenters/ExternalLinkIcon";
import stylesheet from "./stylesheet";

/** @todo Consider extending TextLink */
export default class Link extends Component {
  static propTypes = {
    /** URL to navigate to when clicking this link */
    link: PropTypes.string,
    /** Called when link is clicked  */
    onClick: PropTypes.func,
    /** Called when link is focused  */
    onFocus: PropTypes.func,
    /** Called when hovering over the link */
    onMouseOver: PropTypes.func,
    /** Corresponds to the anchor tag's target */
    target: PropTypes.oneOf(AVAILABLE_TARGETS),
    /** Link text */
    title: PropTypes.string
  };

  static defaultProps = {
    onClick: () => {},
    onMouseOver: () => {}
  };

  _renderExternalLinkIcon = (styles, size) =>
    this.props.target === "_blank" && (
      <div className={css(styles.externalIcon)}>
        <ExternalLinkIcon size={size} />
      </div>
    );

  render() {
    const { title, link, onClick, onFocus, onMouseOver, target } = this.props;
    const Wrapper = link ? "a" : "div";

    return (
      <ThemeContext.Consumer>
        {({ resolvedRoles, metadata }) => {
          const styles = stylesheet(this.props, resolvedRoles);
          const size =
            metadata.densityId === "medium-density"
              ? iconSizes.PX_24
              : iconSizes.PX_16;
          return (
            <Wrapper
              className={css(styles.wrapper)}
              href={link}
              target={target}
              onClick={onClick}
              onFocus={onFocus}
              onMouseOver={onMouseOver}
            >
              <Typography elementType="span" style={styles.typography}>
                {title}
              </Typography>
              {this._renderExternalLinkIcon(styles, size)}
            </Wrapper>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
