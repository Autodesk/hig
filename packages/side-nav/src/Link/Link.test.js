import React from "react";
import renderer from "react-test-renderer";

import Link from "./Link";

describe("side-nav/Link", () => {
  describe("snapshot tests", () => {
    const cases = [
      {
        description: "renders with no props",
        props: {},
      },
    ];

    cases.forEach(({ description, props: { children, ...otherProps } }) => {
      it(description, () => {
        const wrapper = <Link {...otherProps}>{children}</Link>;
        const tree = renderer.create(wrapper).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
