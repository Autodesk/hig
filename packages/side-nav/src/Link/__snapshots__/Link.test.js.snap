// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`side-nav/Link snapshot tests renders with no props 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0 16px;
}

.emotion-1:hover span {
  color: #0696d7;
}

.emotion-1:hover {
  cursor: pointer;
}

.emotion-1:hover svg * {
  fill: #0696d7;
}

.emotion-0 {
  color: #3c3c3c;
  display: block;
  font-family: ArtifaktElement,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.428571429;
  margin: 0;
  text-align: initial;
}

<div
  className="emotion-1"
  href={undefined}
  onClick={[Function]}
  onFocus={undefined}
  onMouseOver={[Function]}
  target={undefined}
>
  <span
    className="emotion-0"
    style={
      Object {
        "lineHeight": "26px",
      }
    }
  />
</div>
`;
